### PR TITLE
Fix preempt: when skipping continue on preempting, the overused judgement didn't consider the current task's resource

### LINKED ISSUE #3161: Preemption fails when queue resources are insufficient and node resources are sufficient
**How to reproduce it (as minimally and precisely as possible)**:

one node has 12 CPU
queue-a capacity is 4 CPU
low-priority-job request 3 CPU
high-priority-job request 3 CPU

1. create low-priority job
2. create high-priority job
3. It can be observed that low-priority job are not preempted

**What you expected to happen**:

Low-priority job are evicted and high-priority job run normally

**Anything else we need to know?**:

preempt will break in here

https://github.com/volcano-sh/volcano/blob/5ab580c83982e4809ca2988552b42aada8ea193c/pkg/scheduler/actions/preempt/preempt.go#L275C34-L275C34

**Environment**:
- Volcano Version: 1.18
- Kubernetes version (use `kubectl version`): 1.27

### PR DESCRIPTION
**Pull request by lowang-bh**
Fix #3161 (Preemption fails when queue resources are insufficient and node resources are sufficient )

### CODE REVIEW

**Comment by lowang-bh** [comment]
@wangyang0616 @hwdef @wpeng102 plz help to review this pr, thanks.

File: pkg/scheduler/util/test_utils.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
 
-	schedulingv2 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+	schedulingv1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
```
Comment (Monokaix) [inline pkg/scheduler/util/test_utils.go:33]: schedulingv1beta1 is better.

Comment (lowang-bh) [reply pkg/scheduler/util/test_utils.go]: Done.

**Comment by hwdef** [comment]
I will check this late, thanks!

**Comment by lowang-bh** [comment]
> Did you test this? In my tests, this patch didn't work, and the higher priority jobs still didn't preempt the lower priority jobs.
> 
> scheduler configmap
> 
> ```yaml
>     actions: "allocate,preempt, backfill"
>     tiers:
>     - plugins:
>       - name: priority
>       - name: gang
>       - name: conformance
>     - plugins:
>       - name: overcommit
>       - name: drf
>         enablePreemptable: false
>       - name: predicates
>       - name: proportion
>       - name: nodeorder
>       - name: binpack
> ```

What does the log show?

**Comment by lowang-bh** [comment]
Hi, @hwdef your case can not work, because volcano webhook will change the minMinber. 
you can test with this config.
### *`minAvailable is 1 and replicas is 3 and each replica request 1cpu.`*
image

**Comment by Monokaix** [comment]
> Did you test this? In my tests, this patch didn't work, and the higher priority jobs still didn't preempt the lower priority jobs.
> 
> scheduler configmap
> 
> ```yaml
>     actions: "allocate,preempt, backfill"
>     tiers:
>     - plugins:
>       - name: priority
>       - name: gang
>       - name: conformance
>     - plugins:
>       - name: overcommit
>       - name: drf
>         enablePreemptable: false
>       - name: predicates
>       - name: proportion
>       - name: nodeorder
>       - name: binpack
> ```

Please add the following param at gang pluing config.
```
        enablePreemptable: false
        enableJobStarving: false
```

**Comment by hwdef** [comment]
> /hold cancel

You shouldn't cancel other people's hold casually.

**Comment by Cdayz** [comment]
Hey, this bug affects out systems too, we are very interested in fixing it

**Comment by lowang-bh** [comment]
> Hey, this bug affects out systems too, we are very interested in fixing it

Hi, @Cdayz , you can use this pr to verify your situation.

File: pkg/scheduler/actions/preempt/preempt.go [hunk 2]
```diff
@@ -271,8 +271,15 @@ func preempt(
...
 			// If reclaimed enough resources, break loop to avoid Sub panic.
-			// If preemptor's queue is overused, it means preemptor can not be allocated. So no need care about the node idle resource
-			if !ssn.Overused(currentQueue) && preemptor.InitResreq.LessEqual(node.FutureIdle(), api.Zero) {
+			// Preempt action is about preempt in same queue, which job is not allocatable in allocate action, due to:
+			// 1. cluster has free resource, but queue not allocatable
+			// 2. cluster has no free resource, but queue not allocatable
+			// 3. cluster has no free resource, but queue allocatable
+			// for case 1 and 2, high priority job/task can preempt low priority job/task in same queue;
+			// for case 3, it need to do reclaim resource from other queue, in reclaim action;
+			// so if current queue is not allocatable(overused about current preemptor) or current idle resource is not enougth for preemptor, it need to continue preempting
```
Comment (Monokaix) [inline pkg/scheduler/actions/preempt/preempt.go:280]: about -> add?

Comment (lowang-bh) [reply pkg/scheduler/actions/preempt/preempt.go:280]: overused about current preemptor: use 'about' to express it is overused when consider current preemptor.

Comment (Monokaix) [reply pkg/scheduler/actions/preempt/preempt.go:280]: Maybe change to "the queue will be overused when consider current preemptor's requests" is better: )

File: pkg/scheduler/actions/preempt/preempt.go [hunk 3]
```diff
@@ -271,8 +271,16 @@ func preempt(
...
+			// Preempt action is about preempt in same queue, which job is not allocatable in allocate action, due to:
+			// 1. cluster has free resource, but queue not allocatable
+			// 2. cluster has no free resource, but queue not allocatable
+			// 3. cluster has no free resource, but queue allocatable
+			// for case 1 and 2, high priority job/task can preempt low priority job/task in same queue;
+			// for case 3, it need to do reclaim resource from other queue, in reclaim action;
+			// so if current queue is not allocatable(the queue will be overused when consider current preemptor's requests)
+			// or current idle resource is not enougth for preemptor, it need to continue preempting
+			// otherwise, break out
+			if ssn.Allocatable(currentQueue, preemptor) && preemptor.InitResreq.LessEqual(node.FutureIdle(), api.Zero) {
```
Comment (wangyang0616) [inline pkg/scheduler/actions/preempt/preempt.go:283]: If the current capacity of the queue is 8C, 5C has been used, and the preemptor request is 4C, and the actual capacity of the queue exceeds 8C, allocatable should return false. The judgment logic here does not have a break, which is not in line with expectations.

I understand that the judgment logic here should be like this:
1. allocatable returns false and breaks directly
2. allocatable returns true, and then judges preemptor and futureIdle. 
    1) preemptor is less than or equal to futureIdle, execute break
    2) preemptor is less than or equal to futureIdle, break is not executed

**Review by william-wang — APPROVED** [review]
/approve

**Pull request merged**